feat: implement student lookup with validation and test coverage

Lokeshloks-1999 opened the pull request:
#  Student Lookup Feature

## Summary

This PR introduces the `findStudentById`  function along with comprehensive test cases to support student lookup functionality based on ID. It enhances the robustness of the student data system by ensuring correct handling of valid, invalid, and edge cases.

---

##  What's Added

- **`findStudentById.ts`**:  function to retrieve a student by their ID from a provided student getter function.
- **Unit tests** for:
  - Valid ID returns correct student.
  - Invalid ID returns `undefined`.
  - Invalid getter (e.g., returns null) throws a descriptive error.

---

##  Test Coverage

- All major paths are tested including:
  - Successful student retrieval
  - ID not found scenario
  - Data source (getter) is invalid scenario

---

## Dependency Note

This PR **imports the `Email` value object**, which was implemented in a previous PR:  
`fix: add Email value object with validation and tests`

## GitHub Issues

- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/434
- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/435

Lokeshloks-1999 committed:
feat(studentData): add findStudentById to retrieve student by ID

Implemented a helper function findStudentById that takes a student ID and a student getter function.
It loads all students and returns the matching student if found. Returns undefined if no match or invalid data.

Lokeshloks-1999 committed:
Add unit test for  to validate student lookup by ID

- Implemented test to ensure correct student is returned when a valid ID is provided.
- Used mock data with an  value object.
 Note:  class was added in the PR titled  by a team member.

Lokeshloks-1999 committed:
Add test to verify  returns undefined for non-existent ID

- Ensures the function handles invalid IDs gracefully by returning undefined.

Lokeshloks-1999 committed:
Add test to ensure  throws error on invalid data from getter

- Validates error handling when the student data getter returns invalid (non-array) data.
- Confirms the function throws the expected error message.

Lokeshloks-1999 committed:
Export  from index for centralized access

- Added  to the main export file to allow usage across modules.

Lokeshloks-1999 committed:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_lookup

Lokeshloks-1999 committed:
feat: Add StudentEmailWrapper to display student email using URL ID and findStudentById

- Extracts student ID from URL using useParams
- Uses findStudentById to locate student from studentGetter list
- Handles missing ID, not found, and error cases gracefully
- Renders StudentEmail component if student is found

Lokeshloks-1999 committed:
Updated the file name from StduentEmailWrapper to StudentEmailWrapper

Lokeshloks-1999 committed:
test: Add unit test for StudentEmailWrapper with valid student ID

- Uses MemoryRouter and Routes to simulate route-based rendering
- Verifies that StudentEmailWrapper correctly renders student name and email

Lokeshloks-1999 committed:
test: Add test for StudentEmailWrapper when student ID is not found

- Simulates routing with unknown student ID using MemoryRouter
- Verifies that the component displays 'Student not found' error message

Lokeshloks-1999 committed:
chore: updated the import name from Email to email

Lokeshloks-1999 commented:
Hello professor,
I’ve implemented the StudentEmailWrapper component, which properly uses findStudentById to load student details. The StudentEmailWrapper acts as the controller logic component for StudentEmail. 
Please let me know if any further adjustments are needed.

Lokeshloks-1999 committed:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_lookup

Lokeshloks-1999 committed:
refactor: rename component to StudentDetailsByID and make it independent to display student details by ID

Lokeshloks-1999 committed:
test: add unit test for StudentDetailsByID rendering student details by ID

Lokeshloks-1999 committed:
feat: add route for StudentDetailsByID to display student details by ID

Lokeshloks-1999 commented:
Hello Professor @achow-conestogac 

I wanted to let you know that I’ve added a route in the application for displaying student details by ID.

Specifically, I created a route at /students/:id that renders the StudentDetailsByID component. This component is now independent and shows the student’s details based on the URL parameter. For now, I’ve set up the studentGetter with empty data as a placeholder, so it can later be connected to the real data source without further routing changes.

This setup is important because in the upcoming milestones, the email component will rely on the same logic to fetch each student’s details dynamically by ID when composing emails. Setting this up early ensures the routing and data retrieval are consistent and will simplify integrating the email functionality later on.

Please let me know if you’d like me to adjust anything.

Lokeshloks-1999 committed:
Refactor : Moved the router of student detail

achow-conestogac commented on src/App.tsx:
you have the same problem as @Shiva30jun and @Charan-devv so i will copy/paste the message here:

this usage does not make sense.

Your visualization is part of enrollment.  it should not be a route that you navigate to.

you need to put your changes into the appropriate place in the Enrollment Dashboard.

achow-conestogac replied:
hello lokesh, you have not address this comment.

I am telling you that you should NOT be using routing but to implement your code directly into the dashboard.

Lokeshloks-1999 replied:
Hello Professor, Thank you for the clarification. I now understand that the expectation is to avoid routing altogether and instead integrate the visualization directly into the Enrollment Dashboard.

I had initially followed the structure from Shiva’s PR(#573), which is why I approached it the same way. In hindsight, I realize this was more of a planning and integration oversight on our part during the initial stages. I did explore different ways to keep the visualization code independent of the StudentList, but due to the tight coupling, it was quite challenging.

Once Shiva’s code is approved, I will align my changes accordingly and refactor the implementation to properly integrate it within the dashboard component, as per your direction.

Lokeshloks-1999 committed:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_lookup

Lokeshloks-1999 committed:
Made sure the current branch is upto date with main

Lokeshloks-1999 committed:
Refactor: nested the StudentDetailsByID route inside the enrollment

Lokeshloks-1999 committed:
Updated the route path to ensure that the branch is building correctly

achow-conestogac commented on src/App.tsx:
```diff
@@ -1,14 +1,17 @@
       <Routes>
-        <Route path="/enrollment" element={<EnrollmentDashboard />} />
+        <Route path="/enrollment" element={<EnrollmentDashboard />} >
+          <Route path="students/id" element={<StudentDetailsByID studentGetter={() => []}/>}/>
```
why does the `studentGetter` do nothing.

achow-conestogac replied:
there's a problem with your structure then.

this is the "routing page".  you should not be trying to get data on this page.

so this means that the `EnrollmentDashboard` should have the data and should be the one giving that to you.

meaning, you should not be defining this as a route.  Rather on the `EnrollmentDashboard`, you need to pass in the correct data when this page is supposed to be displayed.

Lokeshloks-1999 replied:
@achow-conestogac 
Thank you again.

You're right — the routing layer shouldn't manage data logic directly. Initially, I exposed StudentDetailsByID as a route in App.tsx for early development and testing purposes, but you're correct that this doesn't align with best practices.

To clarify, StudentDetailsByID is a presentation component that displays detailed information for a student based on their ID. While it will eventually be reused across modules like StudentEmail, it’s not intended to be a standalone route at this point. Instead, it will be conditionally rendered inside components like StudentEmail, which does belong within the EnrollmentDashboard.

Given that, I’ll:
- Remove the route for StudentDetailsByID from App.tsx
- Ensure that EnrollmentDashboard handles routing to modules like StudentEmail
- Render StudentDetailsByID within StudentEmail using shared context for student data

This should align better with our structure and avoid premature routing setup.

Please let me know if you'd like me to approach this differently.

achow-conestogac commented on src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.test.tsx:
there are more things you can test for.

for one, there's more than 2 elements that you are displaying. 

there are also a bunch of error states that you have not tested.

achow-conestogac commented on src/Enrollment/Student/studentData/findStudentById.test.ts:
```diff
@@ -0,0 +1,40 @@
+import { findStudentById } from "./findStudentById";
+
+import { Student } from "./studentTypes";
+
+
```
fix formatting.

achow-conestogac commented on src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx:
```diff
@@ -0,0 +1,61 @@
+    );
+  }
+
+  let student;
+  try {
+    student = findStudentById(id, studentGetter);
+  } catch (err) {
+    return (
+      <div className="p-8 text-red-600 text-lg">
+        Error: {(err as Error).message}
+      </div>
+    );
+  }
```
dont do this. best practice is to assign the variable as soon as it is used.

your `findStudentById` can return `undefined`. check for that instead.

```suggestion
  const student = findStudentById(id, studentGetter);
  
  if (!student) { // explanation: ! means "not" so if you do undefined is considered false so the opposite of false is true.  so if undefined, then you will end up inside this if block which will return 
    return (
      <div className="p-8 text-red-600 text-lg">
        Error: {(err as Error).message} // you need to fix this error thing here. this never worked as your function never throws an error.  this shows that TDD is not followed.
      </div>
    );
  }
```

Lokeshloks-1999 replied:
I’ve updated the code by removing the unnecessary try-catch block and directly assigning the student variable when it’s needed. I also cleaned up the logic to handle the undefined case without relying on error messages. The component now follows best practices and is much cleaner.

Lokeshloks-1999 committed:
test(StudentDetailsByID): assert rendering of all fields for valid student

Lokeshloks-1999 committed:
test(StudentDetailsByID): add test for missing ID in URL scenario

Lokeshloks-1999 committed:
test(StudentDetailsByID): add test for non-existent student ID

Lokeshloks-1999 committed:
Fixed the formatting and added new import

Lokeshloks-1999 committed:
refactor: simplify StudentDetailsByID by removing unnecessary try-catch and using direct assignment

Lokeshloks-1999 committed:
Updated the app.tsx file

Lokeshloks-1999 committed:
fix(StudentDetailsByID): use string interpolation to correctly display dynamic student ID in not found message

Lokeshloks-1999 committed:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_lookup

Lokeshloks-1999 committed:
Refactor StudentDetailsByID to use local state and loading logic

- Replaced props-based student data retrieval with internal state using useEffect
- Added loading indicator while fetching student details
- Improved conditional rendering for missing ID and not found cases
- Stubbed getAllStudents function for future data source integration

Lokeshloks-1999 committed:
Refactor StudentDetailsByID test to use internal studentGetter logic

- Removed explicit  prop from test cases
- Updated component usage to rely on internal logic
- Ensured all test cases still pass with updated implementation

Lokeshloks-1999 committed:
Refactor routing structure in App component
- Removed unused studentGetter prop from StudentDetailsByID

Lokeshloks-1999 committed:
Updated the Dashboard to show the student detail

Lokeshloks-1999 committed:
Implemented entry point for viewing detailed student information via student table

Lokeshloks-1999 committed:
Updated the import

Lokeshloks-1999 commented on src/Enrollment/Student/Pages/StudentEmail/StudentDetailsByID.tsx:
```diff
@@ -0,0 +1,48 @@
+  if (!student) {
+    return (
+      <div className="p-8 text-red-600 text-lg">
+        Student not found for ID = {id}
```
I have removed the id from the message for privacy and security reasons.

Lokeshloks-1999 commented:
Hello professor, I've updated StudentdetailsbyId to handle all null checks
- When Id not present
- When searching student information, show loading
- When student information not found for the given Id in the route.

I also moved the getterfunction from the StudentdetailbyId route in App.tsx to the component itself, which will be replaced during integration.

Lokeshloks-1999 committed:
Refactor : restructured of studentdetailsbyid, usage of local storage to get students

Lokeshloks-1999 commented on src/Enrollment/Student/Pages/StudentList/StudentList.tsx:
```diff
@@ -54,6 +58,7 @@ const StudentList: React.FC = () => {
             <th className="px-4 py-3 text-right">Actions</th>
           </tr>
         </thead>
+        {/* wire the navigateToStudentDetails to onClick of the student name in the row */}
```
Did the cleanup of comment.

Lokeshloks-1999 commented on src/Enrollment/Student/Pages/StudentList/StudentList.tsx:
Removed our code from studentlist, so no tests are required.

Lokeshloks-1999 commented:
Hello professor, I have updated my code accordingly. Please review it.

Lokeshloks-1999 committed:
Handling null check of student data from cache
